vulnerabilities/data_source.py

```diff
@@ -0,0 +1,371 @@
+        """
+        Subclasses can perform more complex validation than what is handled by data classes and their type annotations.
+
+        This method is called in the constructor. It should raise InvalidConfigurationError with a human-readable
+        message.
+        """
+        pass
+
+    def added_advisories(self) -> Set[Advisory]:
+        """
+        Subclasses yield batch_size sized batches of Advisory objects that have been added to the data source
+        since self.cutoff_date.
+        """
+        raise StopIteration
+
+    def updated_advisories(self) -> Set[Advisory]:
+        """
+        Subclasses yield batch_size sized batches of Advisory objects that have been modified since
+        self.cutoff_date.
+
+        NOTE: Data sources that do not enable detection of changes to existing records vs added records must only
+              implement this method, not added_advisories(). The ImportRunner relies on this contract to decide between
+              insert and update operations.
+        """
+        raise StopIteration
+
+    def error(self, msg: str) -> None:
+        """
+        Helper method for raising InvalidConfigurationError with the class name in the message.
+        """
+        raise InvalidConfigurationError(f'{type(self).__name__}: {msg}')
+
+
+@dataclasses.dataclass
+class GitDataSourceConfiguration(DataSourceConfiguration):
+    repository_url: str
+    branch: Optional[str] = None
+    create_working_directory: bool = True
+    remove_working_directory: bool = True
+    working_directory: Optional[str] = None
+
+
+class GitDataSource(DataSource):
+    CONFIG_CLASS = GitDataSourceConfiguration
+
+    def validate_configuration(self) -> None:
+
+        if not self.config.create_working_directory and self.config.working_directory is None:
+            self.error('"create_working_directory" is not set but "working_directory" is set to the default, which '
+                       'calls tempfile.mkdtemp()')
+
+        if not self.config.create_working_directory and not os.path.exists(self.config.working_directory):
+            self.error('"working_directory" does not contain an existing directory and "create_working_directory" is '
+                       'not set')
+
+        if not self.config.remove_working_directory and self.config.working_directory is None:
+            self.error('"remove_working_directory" is not set and "working_directory" is set to the default, which '
+                       'calls tempfile.mkdtemp()')
+
+    def __enter__(self):
+        self._ensure_working_directory()
+        self._ensure_repository()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.config.remove_working_directory:
+            shutil.rmtree(self.config.working_directory)
+
+    def added_advisories(self) -> Set[Advisory]:
+        raise NotImplementedError
+
+    def updated_advisories(self) -> Set[Advisory]:
+        raise NotImplementedError
+
```

Comment:
How about adding `deleted_advisories` abstract method ? Implementing this method for `GitDataSource`(and other future data sources), will solve the below TODO partially. `ImportRunner`  will also need to have the logic to deal with  deleted advisories.

Reply:
That would be easy enough to do, but I don't know how it would be used by classes inheriting form `DataSource` and I don't want to add methods speculatively. We can iterate and improve on it later. Let me just finish this bloody PR with as little extra fuss as possible. :)

Reply:
Totally fine with me.